pipenv/project.py

```diff
@@ -26,7 +33,13 @@ def __init__(self):
     @property
     def name(self):
         if self._name is None:
-            self._name = self.pipfile_location.split(os.sep)[-2]
+            slug = slugify(self.pipfile_location.split(os.sep)[-2])
+            hash = blake2b(self.pipfile_location.encode(), digest_size=3)
```

Comment:
So two notes here, the first is are we sure we want to have the hash in the project name? This is will change if the directory is ever moved, while the actual project name stays the same. This seems like this dirty laundry doesn't need to be exposed to the user in a public property, especially since it won't function consistently with `PIPENV_VENV_IN_PROJECT`. I'd be in favor of having this functionality moved to a separate function and only used in `virtualenv_location`.

The second issue is, while having a 4 char hash is nicer for readability, I'm consistently encountering a collision with this approach in ~1/5500 project names. As names get longer (10-15 characters), I'm encountering collisions in less than 1/1000. Those aren't great odds for guaranteed uniqueness.